sdk/servicebus/Microsoft.Azure.ServiceBus/src/Primitives/ConcurrentExpiringSet.cs

```diff
@@ -4,64 +4,71 @@
 namespace Microsoft.Azure.ServiceBus.Primitives
 {
     using System;
+    using System.Collections.Generic;
     using System.Collections.Concurrent;
+    using System.Threading;
     using System.Threading.Tasks;
 
     sealed class ConcurrentExpiringSet<TKey>
     {
         readonly ConcurrentDictionary<TKey, DateTime> dictionary;
-        readonly object cleanupSynObject = new object();
-        bool cleanupScheduled;
-        static TimeSpan delayBetweenCleanups = TimeSpan.FromSeconds(30);
+        readonly ICollection<KeyValuePair<TKey, DateTime>> dictionaryAsCollection;
+        CancellationTokenSource tokenSource = new CancellationTokenSource(); // doesn't need to be disposed because it doesn't own a timer
+        volatile TaskCompletionSource<bool> cleanupTaskCompletionSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+        static readonly TimeSpan delayBetweenCleanups = TimeSpan.FromSeconds(30);
 
         public ConcurrentExpiringSet()
         {
             this.dictionary = new ConcurrentDictionary<TKey, DateTime>();
+            this.dictionaryAsCollection = dictionary;
+            _ = CollectExpiredEntriesAsync(tokenSource.Token);
```

Comment:
What happens if there is an exception thrown in CollectExpiredEntriesAsync?

Reply:
@KrzysztofCwalina it would behave the same way as the previous code. It would raise the `TaskUnobservedException` event handler. So far the new implementation only caters for `OperationcanceledException`.  Do you want me to address this in the PR?